### PR TITLE
BUG: complex special case for clip

### LINKED ISSUE #15630: np.clip with complex input is untested and has odd behavior
This came up in https://github.com/pytorch/pytorch/issues/33568. This seems odd:
```
>>> np.clip([3 + 4.j], -1, 2)                                                                                       
array([2.+0.j])
>>> np.clip([3 + 4.j], -1+1.j, 2+12.j)  # imaginary component goes up                                                                   
array([2.+12.j])
>>> np.clip([1 + 4.j], -1+1.j, 2+12.j)  # imaginary component doesn't go up                                                                             
array([1.+4.j])
```

The only test for complex input is that it doesn't segfault. 

Reasonable behavior could be one of:
- Clip real and imaginary parts separately
- Clip absolute value, not changing the phase
There may be other options. As long as it's documented, I'm not sure it matters much which choice is made.

I don't think this is a very important issue, but it would be nice to at least have the desired behavior documented here.

### PR DESCRIPTION
**Pull request by birm**
fixes: #15630

I'm not convinced that this is what's desired, or that the test added is sufficient and/or in the right place. But, it's hopefully a start.

(If someone sees this early, I'll squash my fixup commits once things look passable)

### CODE REVIEW

**Comment by mattip** [comment]
Maybe you could run `python runtests.py` to build and test locally rather than starting ~20 CI runs for each small change

File: numpy/core/src/umath/clip.c.src [hunk 1]
```diff
@@ -31,19 +31,13 @@
...
 #define _NPY_@name@_MAX(a, b) (npy_isnan(a) ? (a) : PyArray_MAX(a, b))
 /**end repeat**/
 
-/**begin repeat
- * #name = CFLOAT, CDOUBLE, CLONGDOUBLE#
- */
-#define _NPY_@name@_MIN(a, b) (npy_isnan((a).real) || npy_isnan((a).imag) || PyArray_CLT(a, b) ? (a) : (b))
-#define _NPY_@name@_MAX(a, b) (npy_isnan((a).real) || npy_isnan((a).imag) || PyArray_CGT(a, b) ? (a) : (b))
-/**end repeat**/
-
```
Comment (mattip) [inline numpy/core/src/umath/clip.c.src]: Aren't these needed elsewhere?

Comment (birm) [reply numpy/core/src/umath/clip.c.src]: they're undefined at the end of the repeat section.

Comment (mattip) [reply numpy/core/src/umath/clip.c.src]: no,sorry, ignore

File: numpy/core/tests/test_multiarray.py [hunk 2]
```diff
@@ -4426,6 +4426,12 @@ def test_nan(self):
         expected = np.array([-1., np.nan, 0.5, 1., 0.25, np.nan])
         assert_array_equal(result, expected)
 
+    def test_complex(self):
+        val = np.array([0+7j, 1+6j, 2+5j, 3+4j])
+        result = val.clip(1+5j, 2+6j)
+        expected = np.array([1+6j, 1+6j, 2+5j, 2+5j])
```
Comment (Qiyu8) [inline numpy/core/tests/test_multiarray.py]: According to the clip definition, values outside the interval are clipped to the interval edges, so the value of expected should be np.array([1+5j, 1+6j, 2+5j, 2+6j]) ?

**Comment by eric-wieser** [comment]
Is this change worth it vs just having the user write:
```
a.real = a.real.clip(...)
a.imag = a.imag.clip(...)
```
It's not clear to me that what you're proposing is a sensible generalization of `clip`

**Comment by birm** [comment]
Honestly, I don't think there _is_ a fully sensible definition of clip for imaginary numbers. Both of the proposals from the linked issue make some sense, but trying to compare values of incomparable values doesn't make much sense to me personally. That said, right now the behavior of clip on complex inputs is just wrong. Would it be preferable to disallow complex inputs?

**Comment by birm** [comment]
I'm going to close this as I don't think that this is even the way to accomplish this, and a maintainer thinks it's a bad idea anyway.

**Pull request closed without merging**